File: leetcode/py/p929-unique-email-addresses.py
# LeetCode 929. Unique email addresses.
from typing import List


class Solution:
    def numUniqueEmails(self, emails: List[str]) -> int:
        unique_emails = set()
        for email in emails:
            local_name, domain_name = email.split("@")
            plus_index = local_name.find("+")
            if plus_index >= 0:
                local_name = local_name[0:plus_index]
            local_name = local_name.replace(".", "")
            unique_emails.add(f"{local_name}@{domain_name}")
        return len(unique_emails)


if __name__ == "__main__":
    solution = Solution()
    tests = (
        ([], 0),
        (["m.y+name@email.com"], 1),
        (
            [
                "test.email+alex@leetcode.com",
                "test.e.mail+bob.cathy@leetcode.com",
                "testemail+david@lee.tcode.com",
            ],
            2,
        ),
        (["a@leetcode.com", "b@leetcode.com", "c@leetcode.com"], 3),
    )
    for test in tests:
        emails = test[0]
        exp = test[1]
        ans = solution.numUniqueEmails(emails)
        if exp != ans:
            print(f"FAIL. emails: {emails}, exp: {exp}, ans: {ans}")

File: leetcode/cpp/p929-unique-email-addresses.cpp
// LeetCode 929.
#include <algorithm>
#include <iostream>
#include <sstream>
#include <unordered_set>
#include <vector>

class Solution {
public:
    int numUniqueEmails(const std::vector<std::string>& emails)
    {
        std::unordered_set<std::string> unique_emails;
        for (auto email : emails) {
            // Get location of '@'.
            const auto at_it = std::find(email.begin(), email.end(), '@');

            // Get end of local_name
            const auto local_name_end = std::find(email.begin(), at_it, '+');

            // Remove any '.' characters and optional '+...' characters.
            email.erase(std::remove(email.begin(), local_name_end, '.'), at_it);

            unique_emails.insert(std::move(email));
        }
        return static_cast<int>(unique_emails.size());
    }
};

template <typename T>
std::string to_string(const std::vector<T>& v)
{
    std::ostringstream oss;
    oss << '{';
    for (const auto e : v) {
        oss << e << ',';
    }
    oss << "\b}";
    return oss.str();
}

int main()
{
    struct TestCase {
        std::vector<std::string> emails;
        int expected;
    };
    TestCase testCases[] = {
        { {}, 0 },
        { { "a@a.com" }, 1 },
        { { "test.email+alex@leetcode.com", "test.e.mail+bob.cathy@leetcode.com", "testemail+david@lee.tcode.com" }, 2 },
        { { "a@leetcode.com", "b@leetcode.com", "c@leetcode.com" }, 3 },
    };
    Solution s;
    for (const auto& tc : testCases) {
        const auto actual = s.numUniqueEmails(tc.emails);
        if (tc.expected != actual) {
            std::cerr << __FILE__ << ":" << __LINE__ << ", FAIL, " << __FUNCTION__
                      << "(emails: " << to_string(tc.emails) << ")"
                      << ", expected: " << tc.expected
                      << ", actual: " << actual
                      << "\n";
        }
    }
    return 0;
}
